app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb

```diff
@@ -1,65 +1,107 @@
+
+  # Creates an event monitor. Used internally by the Runner.
   #
   def initialize(ems)
     @ems = ems
     @collecting_events = false
     @since = nil
   end
 
-  # Start capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should start/keep being captured.
+  #
   def start
     @collecting_events = true
   end
 
-  # Stop capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should stop being captured.
+  #
   def stop
     @collecting_events = false
   end
 
+  # Used internally by the Runner#monitor_events method.
+  #
   def each_batch
     while @collecting_events
-      yield get_events.collect { |e| JSON.parse(e) }
+      yield get_events.collect(&:to_hash)
     end
   end
 
   private
 
+  # Get a list of events that have happened since the most recent event time.
+  #
+  # Because Azure event timestamps are not necessarily stamped in order, an
+  # issue occurs where we could accidentally skip over events that happen
+  # in quick succession. We must therefore begin our query a couple minutes
+  # back from our most recent timestamp, and filter out any duplicates.
+  #
+  # See https://bugzilla.redhat.com/show_bug.cgi?id=1724312 for details.
+  #
   def get_events
-    # Grab only events for the last minute if this is the first poll
-    filter = @since ? "eventTimestamp ge #{@since}" : "eventTimestamp ge #{startup_interval}"
-    fields = 'authorization,description,eventName,eventTimestamp,resourceGroupName,resourceProviderName,resourceId,resourceType'
-    events = connection.list(:filter => filter, :select => fields, :all => true).sort_by(&:event_timestamp)
-
-    # HACK: the Azure Insights API does not support the 'gt' (greater than relational operator)
-    # therefore we have to poll from 1 millisecond past the timestamp of the last event to avoid
-    # gathering the same event more than once.
-    @since = one_ms_from_last_timestamp(events) unless events.empty?
-    events
+    filter = "eventTimestamp ge #{most_recent_time}"
+    connection.list(:filter => filter, :select => SELECT_FIELDS, :all => true)
   end
 
+  # When the appliance first starts, or is restarted, start looking for events
+  # from a fixed, recent point in the past.
+  #
   def startup_interval
-    format_timestamp(1.minute.ago)
+    format_timestamp(2.minutes.ago)
   end
 
-  def one_ms_from_last_timestamp(events)
-    time = Time.at(one_ms_from_last_timestamp_as_f(events)).utc
-    format_timestamp(time)
-  end
+  # Retrieve the most recent Azure event minus 2 minutes, or the startup interval
+  # if no records are found.
+  #
+  # Go back a maximum of 1 hour if the newest record is older than that to avoid
+  # hitting the account's request limits. This should only happen in practice if
+  # the appliance has been down for a while and was restarted.
+  #
+  def most_recent_time
+    result = EventStream.where(source: 'AZURE').maximum(:timestamp)
```

Comment:
Aside from the problem that this hits the database

1. This doesn't scope by ems_id so it'll get _all_ azure events not just for this EMS
2. This is called by get_events _every time_ so it is doing db queries even though after the first call the event catcher has the most up to date info on the events
3. I don't think we need to try to get all events while the event_catcher was down.  If MiqServer shuts down the event_catcher we specifically spin up another one immediately without waiting for the old one to stop specifically for this reason.  And if the appliance was down then the appliance was down its okay to miss things.

What about something like this?
```
EVENT_TIMESTAMP_BUFFER = 2.minutes
attr_accessor :most_recent_time

def initialize(ems)
  ...
  self.most_recent_time = Time.now - EVENT_TIMESTAMP_BUFFER
end

def get_events
  ...
  self.most_recent_time = events.map(&:event_timestamp).sort.last - EVENT_TIMESTAMP_BUFFER
end
```

Reply:
1. Oops, fixed.
2. I'm not sure I follow. Won't a `sort_by` generate another Ruby array internally, every time? The database query, on the other hand, will have its explain plan cached after the first call for 1 or 2 minutes won't it? The explain plan looks pretty zippy to me.
3. Eh, it only goes back an hour at most the way I wrote it there, but if you say we don't need to worry about it then I guess I won't.

Reply:
> Won't a sort_by generate another Ruby array internally, every time?

`events.max_by(&:event_timestamp)` would prevent the new ruby array

> The database query, on the other hand, will have its explain plan cached after the first call for 1 or 2 minutes won't it? The explain plan looks pretty zippy to me.

Anything done in memory is going to be much faster than a database query, but performance isn't even my main concern here.  We're trying to avoid unnecessary database access by provider workers because we want to run them without db access at all in the (hopefully near) future and stuff like this will make that harder.

Reply:
Ok, will update.

Reply:
As an option, why not to store such intermediate values in the cache? We have memcache in use, which is shared same as the database, but should be faster.
```ruby
def last_timestamp_key
  @last_timestamp_key ||= "AzureEventsLastTimestamp/#{ems.id}"
end
...
[Rails.cache.fetch(last_timestamp_key) { 2.minutes.ago }, 1.hour.ago].max
...
Rails.cache.write(last_timestamp_key, last_timestamp)
```

Reply:
I think just an instance variable is sufficient

Reply:
Problem happens when we have worker restarted. That's why instance variable is not sufficient, and we have this database call. However, what I see, is that we have three possible values of timestamp query:
- `1.hour.ago` (max for the previous events timestamp from the database)
- `1.hour.ago...2.minutes.ago` (somewhere in the range, from the database)
- `2.minutes ago` (when worker starts and there is no events in the database)
What difference is here and what is the logic @djberg96 ?

Reply:
> Problem happens when we have worker restarted.

If the worker is restarted while the appliance is up, we spin up a new event_catcher immediately so going back a few minutes will cover the overlap.  We might send some duplicate events but that's why the timestamp fix is so critical because we already handle this case in core.

Reply:
Ok, updated.